### PR TITLE
166 percentile blending

### PR DESCRIPTION
**Pull request by carolinejjones**
Reference issue #166 
Create a plugin that can be used to blend percentile data. This is an extension of the existing weighted blending plugin rather than a new plugin.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

CLA
 - [x] If a new developer, signed up to CLA

### CODE REVIEW

File: lib/improver/tests/weighted_blend/test_blending_utilities.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+                                                         coord_dim,
+                                                         perc_coord)
+        expected_result_array = np.reshape(BLENDED_PERCENTILE_DATA2,
+                                           (6, 2, 2))
+        self.assertArrayAlmostEqual(result.data, expected_result_array)
+
+    def test_blend_percentile_aggregate(self):
+        """Test blend_percentile_aggregate function works"""
+        weights = np.array([0.8, 0.2])
+        percentiles = np.array([0, 20, 40, 60, 80, 100])
```
Comment (benfitzpatrick) [inline lib/improver/tests/weighted_blend/test_blending_utilities.py]: For a more uneven choice of weights, tends correctly to the high weight distribution except for 0 and 100.

File: lib/improver/weighted_blend.py [hunk 2]
```diff
@@ -33,13 +33,217 @@
...
+        Returns:
+            result : iris.cube.Cube
+                     containing the weighted percentile blend
+                     across the chosen coord
+        """
+
+        percentiles = np.array(perc_coord.points, dtype=float)
+        num = cube.data.shape[coord_dim[0]]
+        if weights is None:
+            weights = np.ones(num)/float(num)
```
Comment (carolinejjones) [inline lib/improver/weighted_blend.py]: Added

File: lib/improver/weighted_blend.py [hunk 3]
```diff
@@ -33,13 +33,217 @@
...
+
+        result = cube.collapsed(coord,
+                                PERCENTILE_BLEND,
+                                arr_percent=percentiles,
+                                arr_weights=weights,
+                                perc_dim=perc_dim)
+
+        # Add meta data
+        new_name = 'over {}'.format(coord)
+        result.attributes.update({'Blended': new_name})
```
Comment (carolinejjones) [inline lib/improver/weighted_blend.py]: No, the document on CF only came out yesterday. It is really just a place holder. I'll have a look in the document and see if I can find something more appropriate.

File: lib/improver/tests/weighted_blend/test_blending_utilities.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+import iris
+from iris.coords import AuxCoord, DimCoord
+from iris.cube import Cube
+from iris.tests import IrisTest
+import numpy as np
+
+from improver.weighted_blend import BlendingUtilities
+
+PERCENTILE_DATA = np.array([
+    15.3077946, 14.65380361, 15.91478244, 15.10887522,
```
Comment (carolinejjones) [inline lib/improver/tests/weighted_blend/test_blending_utilities.py]: Random sample from a Normal distribution. There were taken from the blending notebook as I was happy that the processing in the notebook was working correctly.

File: lib/improver/tests/weighted_blend/test_blending_utilities.py (hunk 1 above)
Comment (carolinejjones) [inline lib/improver/tests/weighted_blend/test_blending_utilities.py]: Yes, 0 and 100 are the range of the combined percentile regardless of how much weight you apply.

Comment (carolinejjones) [inline lib/improver/tests/weighted_blend/test_blending_utilities.py]: As discussed, the plugin only accepts a cube but there is a more general task to reduce incoming data onto a cube. #209 has been raised for this work.

File: lib/improver/weighted_blend.py [hunk 5]
```diff
@@ -33,13 +33,217 @@
...
+        if arr_percent.shape > (1,):
+            shape = arr_percent.shape + shape
+            result = result.reshape(shape)
+            if axis < perc_dim:
+                if perc_dim != 1:
+                    result = np.rollaxis(result, 0, start=perc_dim-1)
+            else:
+                if perc_dim != 0:
+                    result = np.rollaxis(result, 0, start=perc_dim)
+        # print 'Result shape',result.shape
```
Comment (bayliffe) [inline lib/improver/weighted_blend.py]: I think Ben is simply requesting the removal of the debug print statement.

File: lib/improver/weighted_blend.py (hunk 3 above)
Comment (bayliffe) [inline lib/improver/weighted_blend.py]: We will need to form a consensus on this. Ideally the management team will give us a steer on the direction they'd like us to take (?).

Comment (carolinejjones) [inline lib/improver/weighted_blend.py]: For now I have taken this meta data out as it wasn't obvious from the CF document. There is still a new cell method percentile_blend which iris adds but which isn't part of the CF standard. I'm going to leave it in for now and have updated our internal documents to show that it has been added.

**Comment by bayliffe** [comment]
Replaced by PR #220

**Pull request closed without merging**